File: src/main/kotlin/predcompiler/compilation/io/CSVReader.java
package predcompiler.compilation.io;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class CSVReader {

    public static class CSVResult {
        public String[] headers;
        public float[][] rows;

        public CSVResult(String[] headers, float[][] rows) {
            this.headers = headers;
            this.rows = rows;
        }
    }

    public static CSVResult readCSV(String filePath) throws IOException {
        List<float[]> rowList = new ArrayList<>();
        String[] headers = null;
        
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            // Read the header line
            String headerLine = br.readLine();
            if (headerLine == null) {
                throw new IOException("The file is empty");
            }

            // Split the header line to get column names
            headers = headerLine.split(",");

            // Read the rest of the file line by line
            String line;
            while ((line = br.readLine()) != null) {
                String[] values = line.split(",");
                if (values.length != headers.length) {
                    throw new IOException("Mismatch between number of columns and number of values in a row");
                }
                
                // Parse and store the float values in an array
                float[] rowValues = new float[values.length];
                for (int i = 0; i < values.length; i++) {
                    rowValues[i] = Float.parseFloat(values[i]);
                }
                rowList.add(rowValues);
            }
        }

        // Convert the list of rows to a 2D array
        float[][] rows = new float[rowList.size()][];
        rows = rowList.toArray(rows);

        return new CSVResult(headers, rows);
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java CSVReader <csv-file-path>");
            return;
        }

        String filePath = args[0];
        try {
            CSVResult result = readCSV(filePath);
            System.out.println("Headers: " + arrayToString(result.headers));
            for (float[] row : result.rows) {
                System.out.println("Row: " + arrayToString(row));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static String arrayToString(String[] array) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < array.length; i++) {
            sb.append(array[i]);
            if (i < array.length - 1) {
                sb.append(", ");
            }
        }
        sb.append("]");
        return sb.toString();
    }

    public static String arrayToString(float[] array) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < array.length; i++) {
            sb.append(array[i]);
            if (i < array.length - 1) {
                sb.append(", ");
            }
        }
        sb.append("]");
        return sb.toString();
    }
}

File: src/main/kotlin/predcompiler/compilation/io/CSVReader.kt
package predcompiler.compilation.io

import java.io.File
import java.nio.file.FileVisitResult
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.SimpleFileVisitor
import java.nio.file.attribute.BasicFileAttributes
import kotlin.io.path.pathString

data class CSVResult(
    val headers: List<String>,
    val rows: List<List<Double>>
)

/**
 * Reads a CSV file and returns the headers and rows as a [CSVResult] object.
 * @param filePath the path to the CSV file
 */
fun readCSV(filePath: String): CSVResult {

    File(filePath).bufferedReader().use { br ->

        // Read the header line
        val headerLine = br.readLine()
        require(!headerLine.isNullOrEmpty()) { "The file is empty" }

        // Split the header line to get column names
        val headers = headerLine.split(",")

        // Read the rest of the file line by line
        val rowList = mutableListOf<List<Double>>()
        var line: String
        while ((br.readLine().also { line = it }) != null) {
            val values = line.split(",").toTypedArray()
            check(values.size == headers.size) {
                "Mismatch between number of columns (${headers.size}) and number of values in a row (${values.size})"
            }

            // Parse and store the float values in an array
            rowList.add(values.map { it.toDouble() })
        }
        return CSVResult(headers, rowList)
    } // File(filePath).bufferedReader().use
} // readCSV

/**
 * Collects all CSV files from the specified directory and returns a list of their paths.
 * Provided path must be a valid directory.
 */
fun collectCsvFilesFromPath(dir: Path): List<Path> {
    require(Files.exists(dir) && Files.isDirectory(dir)) { "Directory does not exist: $dir" }

    val csvFiles = mutableListOf<Path>()
    Files.walkFileTree(dir, object : SimpleFileVisitor<Path>() {
        override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
            if (file.toString().endsWith(".csv")) {
                csvFiles.add(file)
            }
            return FileVisitResult.CONTINUE
        }
    })
    return csvFiles
} // collectCsvFilesFromPath

fun readCSVFilesFromPath(path: Path): List<CSVResult> {
    return collectCsvFilesFromPath(path).map { readCSV(it.pathString) }
} // readCSVFilesFromPath

private fun <T> arrayToString(array: List<T>): String {
    val sb = StringBuilder()
    sb.append("[")
    for (i in array.indices) {
        sb.append(array[i])
        if (i < array.size - 1) {
            sb.append(", ")
        }
    }
    sb.append("]")
    return sb.toString()
} // arrayToString

fun main(args: Array<String>) {
    if (args.size != 1) {
        println("Usage: java CSVReader <csv-file-path>")
        return
    }

    val filePath = args[0]
    readCSV(filePath).also { (headers, rows) ->
        println("Headers: " + arrayToString(headers))
        rows.forEach { row ->
            println("Row: " + arrayToString(row))
        }
    }
} // main
